### ISSUE TITLE
Proxy documentation confusing

### ISSUE THREAD

**Issue by urzds** [coreos/etcd#3740]
I am trying to follow the [proxy documentation](https://coreos.com/etcd/docs/latest/proxy.html) to setup an etcd2 cluster of three main nodes `infra01` to `infra03` and make all other machines proxy requests to them.

The resulting `20-cloudinit.conf` for the machines proxying looks like this, which I think conforms to what the documentation tells me to do:
```systemd
[Service]
Environment="ETCD_INITIAL_CLUSTER=infra01=https://192.168.10.24:2380,infra02=https://192.168.10.25:2380,infra03=https://192.168.10.26:2380"
Environment="ETCD_LISTEN_CLIENT_URLS=http://localhost:2379,http://localhost:4001"
Environment="ETCD_NAME=node32"
Environment="ETCD_PROXY=on"
```

etcd2 on those nodes logs:
```
could not get cluster response from http://localhost:2380: Get http://localhost:2380/members: dial tcp 127.0.0.1:2380: getsockopt: connection refused
could not get cluster response from http://localhost:7001: Get http://localhost:7001/members: dial tcp 127.0.0.1:7001: getsockopt: connection refused
proxy: could not retrieve cluster information from the given urls
proxy: zero endpoints currently available
```

What is particular confusing in the documentation is that you first mention the `-advertise-client-urls` flag of the *members* as being authoritative, telling the proxy how to connect to the cluster. (I assume that the *members* are `infraNN` in my case.) Later you mention `--initial-cluster` as the flag that determines how the proxy discovers the nodes of the cluster. So the proxy first connects to the peer-urls to determine the client-urls, and then connects to the client-urls to do it's actual work?

In my specific setup, the *members* of the cluster secure their communication using TLS and certificate verification:
```systemd
[Service]
Environment="ETCD_ADVERTISE_CLIENT_URLS=http://192.168.10.24:2379"
Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=https://192.168.10.24:2380"
Environment="ETCD_INITIAL_CLUSTER=infra01=https://192.168.10.24:2380,infra02=https://192.168.10.25:2380,infra03=https://192.168.10.26:2380"
Environment="ETCD_INITIAL_CLUSTER_STATE=existing"
Environment="ETCD_INITIAL_CLUSTER_TOKEN=test"
Environment="ETCD_LISTEN_CLIENT_URLS=http://localhost:2379,http://localhost:4001,http://192.168.10.24:2379,http://192.168.10.24:4001"
Environment="ETCD_LISTEN_PEER_URLS=https://localhost:2380,https://localhost:7001,https://192.168.10.24:2380,https://192.168.10.24:7001"
Environment="ETCD_NAME=infra01"
Environment="ETCD_PEER_CERT_FILE=/srv/etcd/peer.crt"
Environment="ETCD_PEER_KEY_FILE=/srv/etcd/peer.key"
Environment="ETCD_PEER_CLIENT_CERT_AUTH=true"
Environment="ETCD_PEER_TRUSTED_CA_FILE=/srv/etcd/ca.crt"
```

So it is not really clear how the proxy could connect to the `--initial-cluster`, if it does not have a trusted certificate... In fact it does not show this behaviour at all - instead it seems to directly try to connect to `http://localhost:2379`, instead of any of the machines listed in `--initial-cluster` (neither it nor the `infraNN` nodes mention any failed connection attempts).

1. Could you please clarify on how `--advertise-client-urls` and `--initial-cluster` interact? And which flag is needed on which node?

2. Further, where does it get the `http://localhost:2380` URLs from? It seems the only mention of these URLs are in the `--initial-advertise-peer-urls` or `--initial-cluster` default values. Does etcd in proxy mode use those instead of the actual flags that I provide in `cloud-config.yml` (see the `20-cloudinit.conf` above)?

infra nodes:
```sh
cat /etc/lsb-release 
DISTRIB_ID=CoreOS
DISTRIB_RELEASE=835.1.0
DISTRIB_CODENAME="Red Dog"
DISTRIB_DESCRIPTION="CoreOS 835.1.0"
```

proxy nodes:
```sh
# cat /etc/lsb-release 
DISTRIB_ID=CoreOS
DISTRIB_RELEASE=833.0.0
DISTRIB_CODENAME="Red Dog"
DISTRIB_DESCRIPTION="CoreOS 833.0.0"
```

**Comment by gyuho** [comment]
When you say `--initial-advertise-client-urls`, do you mean `--advertise-client-urls`?
`--initial-advertise-client-urls` does not exist.

**Comment by gyuho** [comment]
For the second question, I think `etcdmain` package defaults `listen-peer-urls` as `"http://localhost:2380,http://localhost:7001"`, if not specified in the flags. `listen-peer-urls` are `"List of URLs to listen on for peer traffic"` [1].

---
[1] https://github.com/coreos/etcd/blob/master/etcdmain/config.go#L156

**Comment by urzds** [comment]
@gyuho That's exactly what I mean. The flag does not exist. Yet the documentation references it. I [quote literally](https://coreos.com/etcd/docs/latest/proxy.html):
```
The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each members' -advertise-client-urls flag.
```

Regarding the 2nd answer: Does that mean that etcd2 in proxy mode tries to connect to `listen-peer-urls`, ignoring both the magical `advertise-client-urls` and `intial-cluster`, which are mentioned in the documentation as the urls it would connect to?

**Comment by gyuho** [comment]
For the first question, as far as I understand, you need `initial-cluster` for initial cluster bootstrapping. (e.g. `-initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380'` for 3-machine cluster). This used to be called `peers` in flags. These are needed for inter-server communication. You need `advertise-client-urls` to `advertise its client URLs to the rest of the cluster`, which are needed to receive requests from clients. So I think you need this configuration on every node.

https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L387 has the line on how proxy gets peer URLs (clusters), if I understand your question correctly.

**Comment by urzds** [comment]
1. First question
  1. So the `initial-cluster` flag is not needed in proxy node, because that flag is only used in initial bootstrapping? Thus the documentation for the proxy mode is wrong in this regard?
  2. It seems I mixed up `initial-advertise-peer-urls` and `advertise-client-urls`. I fixed that in the original post and will try again.

Regarding the code reference you posted: It seems that peerURLs in my case comes from `discovery.GetCluster()` and is then used to call `etcdserver.GetClusterFromRemotePeers()`?

**Comment by raoofm** [comment]
@urzds 

1. i. 
-initial-cluster flag is needed in proxy mode. It should be same as what you used for your etcd cluster

ETCD_INITIAL_CLUSTER="infra01=https://192.168.10.24:2380,infra02=https://192.168.10.25:2380,infra03=https://192.168.10.26:2380"

It uses either -initial-cluster or -discovery or -discovery-srv flag to bootstrap itself(proxy).

   ii.
-listen-client-urls flag is required by proxy and should have something like http://localhost:2377

Proxy doesn't need any other urls. Please delete the data-dir and restart proxy, it should work.

Regarding your confusion with -advertise-client-urls flag. Documentation doesn't say that you need this flag to start proxy. They try to say that set this flag correctly while creating each node of the etcd cluster because if you set it incorrectly (to something like localhost:2379) then the proxy uses this address to communicate with the cluster node and the address localhost is proxy's machine for proxy. So it will not work.

Hope it clarifies!

**Comment by urzds** [comment]
Thanks for the clarification. Especially the hint that I have to delete the data dir - suggesting that the commandline flags are not authoritative.

One more issue, though: Is it correct, that the proxies will connect to the peer urls to determine the client urls, which they will then proxy to? Thus the proxies will need valid peer certificates to determine the (on my installation) not-secured/http client urls? Thus there is no way to split the cluster into peers and clients/proxies by means of certificates?

**Comment by raoofm** [comment]
@urzds yes they need peer urls, to connect to and get client urls to which they can proxy to.

They need not need valid certificates, they just need to have the ca that issued the peer certs to be trusted. It will work without specifying any cert flags if ca is trusted (in case you are using self signed ca then you have to append that to your trusted ca), 

The above is true if you are using https only but if you are enabling client-cert-auth then you will have to set the cert, key and trusted ca as well.

I think this really explains well
https://github.com/coreos/etcd/blob/master/Documentation/security.md#basic-setup

**Comment by urzds** [comment]
Thanks for the clarification. I read the security documentation, but it does not explain how encryption/authentication interacts with proxy mode. Neither does the proxy documentation.

**Comment by raoofm** [comment]
I agree we don't have a doc for proxy.

For me a proxy fits well in the documentation for client.

https://github.com/coreos/etcd/blob/master/Documentation/security.md#example-1-client-to-server-transport-security-with-https
https://github.com/coreos/etcd/blob/master/Documentation/security.md#example-2-client-to-server-authentication-with-https-client-certificates

**Comment by raoofm** [comment]
@urzds where you able to bring up proxies?

**Comment by urzds** [comment]
Yes, after giving the proxies full access by creating trusted certificates for them, they were able to connect to the cluster.

**Comment by bobelev** [comment]
@gyuho your documentation changes made proxy much more understandable. Thank you!

**Comment by gyuho** [comment]
@bobelev No problem. Thanks to @joshix

**Comment by joshix** [comment]
I believe #3871 closes this. Interested if @urzds, @gyuho, @xiang90 and other interested parties agree. @bobelev seems to think this one is closed. :)

**Comment by xiang90** [comment]
@joshix @urzds 

The doc helps. But ideally, the proxy SHOULD NOT talk to etcd cluster with peer urls. It should only reach the client urls. It is very tricky right now due to the etcd member fail-over (we want the proxy and member share the same flag. The flag is peer url right now...)

**Comment by joshix** [comment]
@xiang90 Ok, so if I understand correctly, complete resolution of this issue is blocked pending changes to etcd peer-to-proxy failover?

**Comment by urzds** [comment]
First: Thanks for the credits! It was a pleasure assisting @gyuho in clarifying significant parts of the document.

@joshix @xiang90 One bigger problem in understanding how this works, was to figure out that not only peers connect to peer URLs and that they do not necessarily have to authenticate themselves using a client certificate (or do they...?). I think this is still not explained anywhere. However, even better than a detailed explanation may be to rethink the behaviour, because the process itself would become less confusing, needing less documentation.

P.S: I found a minor typo in the document: "automtically"

**Comment by xiang90** [comment]
Fixed via #3937
